pkg/proxy/cache.go

```diff
@@ -1,22 +1,78 @@
 package proxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/codeready-toolchain/registration-service/pkg/application"
+	"github.com/codeready-toolchain/registration-service/pkg/log"
 	"github.com/codeready-toolchain/registration-service/pkg/proxy/namespace"
+
 	"github.com/gin-gonic/gin"
 )
 
 type UserNamespaces struct {
+	sync.RWMutex
 	app application.Application
+
+	// namespace accesses by username
+	namespaces map[string]*namespace.NamespaceAccess
 }
 
 func NewUserNamespaces(app application.Application) *UserNamespaces {
 	return &UserNamespaces{
-		app: app,
+		app:        app,
+		namespaces: make(map[string]*namespace.NamespaceAccess),
+	}
+}
+
+// Get tries to retrieve the namespace access from the cache first. If found then it checks if the cached access is still valid.
+// If not found or invalid then retrieves a new namespace access from the member cluster service and stores it in the cache.
+func (c *UserNamespaces) Get(ctx *gin.Context, userID, username string) (*namespace.NamespaceAccess, error) {
+	na, err := c.namespaceFromCache(ctx, username)
+	if err != nil {
+		return nil, err
+	}
+	if na != nil {
+		return na, nil
+	}
+
+	// Not found in cache or invalid. Get a new namespace access.
+	log.Info(ctx, fmt.Sprintf("Retrieving a new Namespace Access from the cluster for the user : %s", username))
+	na, err = c.app.MemberClusterService().GetNamespace(ctx, userID, username)
+	if err != nil {
+		return nil, err
+	}
+	// Store in the cache
+	c.add(username, na)
+
+	return na, nil
+}
+
+// namespaceFromCache tries to retrieve a namespace access from the cache and validates it.
+// Returns nil, nil if no namespace access found or if it's invalid.
+func (c *UserNamespaces) namespaceFromCache(ctx *gin.Context, username string) (*namespace.NamespaceAccess, error) {
+	c.RLock()
+	defer c.RUnlock()
+	na, ok := c.namespaces[username]
+	if ok {
+		valid, err := na.Validate()
+		if err != nil {
+			return nil, err
+		}
+		if valid {
```

Comment:
if `na` not valid, shouldn't you remove it from the cache?

Reply:
Well... we could but if the `na` is not valid than we get a new one from the cluster and put it in the cache. Removing it from cache, retrieving a new one and then putting back could actually decrease performance in case of multiple parallel requests from the same user. One go routine may be removing the invalid na which was just updated by another. It wouldn't cause any errors or something like this. Eventually the cache would be properly updated but my point is that it's probably doesn't give us any performance benefits if we remove it before updating.

Reply:
sure, the invalid entry will be replaced by a valid one shortly after. Except maybe when users disconnect from the platform, in which case the cache will retain the entries indefinitely?

Reply:
that being said, clearing outdated entries in the cache could be addressed later

Reply:
Not indefinitely. Until the next pod restart :)

Reply:
actually, it's a separate use-case 🤦‍♂️

Reply:
And it wouldn't actually help much with outdated cache entries for disconnected users anyway. We check if the na is valid only if there is a request from a client/user. It's unlikely that a user which was removed would keep using the proxy.

Reply:
yes, which is why I wrote it's a separate use-case: the code above is only going to be called by active users! ;)